solutions/silver/usaco-1182.mdx

```diff
@@ -0,0 +1,139 @@
+---
+id: usaco-1182
+source: USACO Silver 2022 January
+title: Searching for Soulmates
+author: Chongtian Ma, Chuyang Wang
+---
+
+[Official Analysis (Java)](http://www.usaco.org/current/data/sol_prob1_silver_jan22.html)
+
+
+## Explanation
+
+To transform $a$ into $b$, we proceed greedily. If $a > b$, it can be shown that
+it is never optimal to add more than once consecutively before dividing $a$ by
+2. Then, if $a < b$, we can simulate our strategy above by allowing
+multiplication, division, and subtraction by one on $b$. We therefore either
+divide $b$ by two (and first subtract by one if $b$ is odd) or subtract $b$ $b -
+a$ times to directly reach $a$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log{\max(A, B)})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+
+ll solve(ll a, ll b) {
+	if (a == b) {
+		return 0;
+	} else if (a > b) {
+		/*
+		 * Divide a greedily until a <= b, add 1 in case a is odd to enable
+		 * division.
+		 */
+		ll is_odd = a % 2;
+		return 1 + is_odd + solve((a + is_odd) / 2, b);
+	} else {
+		/*
+		 * Otherwise, simulate multiplication of a by dividing b.
+		 * Alternatively, if it's better, add up to b directly.
+		 */
+		ll is_odd = b % 2;
+		return min(b - a, 1 + is_odd + solve(a, b / 2));
+	}
+}
+
+int main() {
+	int N;
+	cin >> N;
+	for (int i = 0; i < N; i++) {
+		ll a, b;
+		cin >> a >> b;
+		cout << solve(a, b) << endl;
+	}
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class solution {
+	public static long solve(long a, long b) {
+		if (a == b) {
+			return 0;
+		} else if (a > b) {
+			/*
+			 * Divide a greedily until a <= b, add 1 in case a is odd to enable
+			 * division.
+			 */
+			long isOdd = a % 2;
+			return 1 + isOdd + solve((a + isOdd) / 2, b);
+		} else {
+			/*
+			 * Otherwise, simulate multiplication of a by dividing b.
+			 * Alternatively, if it's better, add up to b directly.
```

Comment:
why do we have to *simulate* multiplication of a?

Reply:
~~cause i copied envy's code~~
cause we're only allowed to modify a. What we do on b can be seen as operations on a in reversed order.

Reply:
can you do that op on a to make it more explicit

Reply:
tbh im not sure if we can do it directly on `a` without making the recursion very ugly

Reply:
but looking at it we should be able to do this in just 1 step if b is even & 2 is b is odd?

Reply:
can u elaborate, esp. how to determine how many times u add at the beginning if u work on a?

Reply:
wait what
is this question meant for me

Reply:
> we should be able to do this in just 1 step if b is even & 2 is b is odd?

emm what